pynhm/base/budget.py

```diff
@@ -0,0 +1,172 @@
+        )
+
+        # keep only the desried metadata for budget
+        meta_keys = Budget.get_meta_keys()
+        for component in self.components:
+            for key, val in kwargs["meta"][component].items():
+                kwargs["meta"][component][key] = {
+                    kk: vv for kk, vv in val.items() if kk in meta_keys
+                }
+
+        for component in self.components:
+            kwargs[component] = {}
+            for var in kwargs["meta"][component].keys():
+                kwargs[component][var] = storage_unit[var][:]
+
+        budget = Budget(**kwargs)
+        budget.storage_unit_name = storage_unit.name
+        return budget
+
+    @staticmethod
+    def get_meta_keys():
+        """Return a tuple of the metadata keys used by Budget."""
+        return ("desc", "modules", "var_category", "units")
+
+    @staticmethod
+    def get_components():
+        return ("inputs", "outputs", "storage_changes")
+
+    @property
+    def components(self):
+        return self.get_components()
+
+    def set_initial_accumulations(self, init_accumulations):
+        # init to zero
+        self._accumulations = {}
+        for component in self.components:
+            self._accumulations[component] = {}
+            for var in self[component].keys():
+                self._accumulations[component][var] = (
+                    self[component][var] * zero
+                )
+
+        if init_accumulations is None:
+            return None
+
+        for component in self.components:
+            if component not in init_accumulations.keys():
+                continue
+            for var in self[component].keys():
+                if var in init_accumulations[component].keys():
+                    self._accumulations[component][var] = init_accumulations[
+                        "component"
+                    ][var]
+
+        return None
+
+    def accumulate(self):
+        for component in self.components:
+            for var in self[component].keys():
+                self._accumulations[component][var] += self[component][var]
+        return None
+
+    @property
+    def accumulations(self):
+        return self._accumulations
+
+    def _sum(self, attr):
+        key0 = list(self[attr].keys())[0]
+        sum = self[attr][key0] * zero
+        for kk, vv in self[attr].items():
+            sum += vv
+        return sum
+
+    def _sum_inputs(self):
+        return self._sum("inputs")
+
+    def _sum_outputs(self):
+        return self._sum("outputs")
+
+    def _sum_storage_changes(self):
+        return self._sum("storage_changes")
+
+    def _calc_balance(self):
+        balance = self._inputs_sum - self._outputs_sum
+        close = np.isclose(balance, self._storage_changes_sum)
+        # diff = balance, self._storage_changes_sum)
+        if not close.all():
+            msg = "The flux balance not equal to the change in storage"
+            raise ValueError(msg)
```

Comment:
We probably don't want to die here.  We should be able to calculate a budget and see that it doesn't balance without the script terminating.

Reply:
Agree. I was being a bit dramatic. Maybe that could be an option from the controller?

Reply:
Now
`imbalance_fatal=True` is optional upon instantiation of Budget.
Or class with a budget takes `budget_type='strict'`. 

To be reviewed. Controller could also set inside a class with a budget.